lib/flight_job/render_context.rb

```diff
@@ -0,0 +1,87 @@
+#==============================================================================
+# Copyright (C) 2021-present Alces Flight Ltd.
+#
+# This file is part of Flight Job.
+#
+# This program and the accompanying materials are made available under
+# the terms of the Eclipse Public License 2.0 which is available at
+# <https://www.eclipse.org/legal/epl-2.0>, or alternative license
+# terms made available by Alces Flight Ltd - please direct inquiries
+# about licensing to licensing@alces-flight.com.
+#
+# Flight Job is distributed in the hope that it will be useful, but
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with Flight Job. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on Flight Job, please visit:
+# https://github.com/openflighthpc/flight-job
+#==============================================================================
+
+module FlightJob
+  class RenderContext
+    class AnswerDecorator
+      def initialize(question:, answer:)
+        @question = question
+        @answer = answer
+      end
+
+      def answer
+        @answer || @question.default
+      end
+
+      def default
+        @question.default
+      end
+    end
+
+    def initialize(template:, answers:)
+      @template = template
+      @answers = answers
+    end
+
+    def render
+      @template.to_erb.result(binding)
+    end
+
+    def question
+      questions
+    end
+
+    def questions
+      @questions ||= begin
+        questions = @template.generation_questions.reduce({}) do |memo, question|
+          memo.merge({
+            question.id => AnswerDecorator.new(question: question,
+                                               answer: @answers[question.id])
+          })
+        end
+        DefaultsOpenStruct.new(questions) do |h, k|
+          question = Question.new(id: k)
+          h[k] = AnswerDecorator.new(question: question, answer: nil)
+        end
+      end
+    end
+  end
+
+  class DefaultsOpenStruct < OpenStruct
+    def initialize(opts = {}, &b)
+      if b
+        @table = Hash.new(&b)
+      else
+        @table = {}
+      end
+      @table.define_singleton_method(:key?) { |_| true }
+      opts.each do |k, v|
+        @table[k.to_sym] = v
+      end
+    end
+  end
```

Comment:
What's this class for?  In particular, why is `key?` being defined to always return `true`?

Reply:
The `DefaultOpenStruct` provides some syntactic sugar to the rendering. In particular it provides handle for missing questions. For example:

```
# If a standard OpenStruct is used
questions.missing               # => nil
questions.missing.answer  # => NoMethodError

# When using the DefaultOpenStruct
question.missing               # => AnswerDecorator
question.missing.answer  # => nil
```

The `OpenStruct` object does not support the defaults mechanism provided by `Hash`. It also duplicates the input `Hash`, which prevents defaults being support via the original hash.

Also because there are infinitely many "missing keys", the standard `defaults_hash.merge(answers)` mechanism does not work.

So instead the `DefaultsOpenStruct` bridges the gap between the two. It is an `OpenStruct` that supports the defaults mechanism provided by `Hash`.

In regards to the `key?` method, because it is an "OpenStruct" with defaults, it responds to all keys. Thus the `key?` method should always be `true`.

Reply:
Thanks for the response.  Could you add a brief comment to this affect to the class.

I wonder if we'll eventually want to differentiate between a question existing or not.  If we do, we might want to change the behaviour of `key?`.  But let's leave that for future us to worry about.

Reply:
For future reference,

Changing the behaviour of `key?` whilst still inheriting off `OpenStruct` is probably a bad idea. I believe that `OpenStruct` uses `key?` internally so I'm not sure what side effects it will introduce.

In general this is a quick and dirty way to implement the rendering features we want. There is probably scope for replacing it with a dedicated object at some point in the future.

Reply:
19dde4487630fad7b6bec3d862690badd39e724b